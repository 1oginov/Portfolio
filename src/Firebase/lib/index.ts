import * as firebase from 'firebase/app';
import 'firebase/auth';

import * as C from 'Constants';

export interface GetFirebaseExtraArgument {
  (): firebase.app.App;
}

export interface WithFirebaseHocProps {
  firebase: firebase.app.App;
}

export const getAuthProvider = (provider: string): firebase.auth.AuthProvider => {
  switch (provider) {
    case 'google':
      return new firebase.auth.GoogleAuthProvider();

    default:
      throw new Error('Unknown auth provider');
  }
};

export const getPositionsCollectionPath = (userId: string): string => (
  `${C.FIRESTORE_USERS_COLLECTION}/${userId}/${C.FIRESTORE_POSITIONS_COLLECTION}`
);

export const getPositionDocumentPath = (userId: string, positionId: string): string => (
  `${C.FIRESTORE_USERS_COLLECTION}/${userId}/${C.FIRESTORE_POSITIONS_COLLECTION}/${positionId}`
);

export const getPositionsCollection = (
  firebaseInstance: firebase.app.App,
  userId: string,
): firebase.firestore.CollectionReference => (
  firebaseInstance.firestore()
    .collection(C.FIRESTORE_USERS_COLLECTION)
    .doc(userId)
    .collection(C.FIRESTORE_POSITIONS_COLLECTION)
);

export const getPositionDocument = (
  firebaseInstance: firebase.app.App,
  userId: string,
  positionId: string,
): firebase.firestore.DocumentReference => getPositionsCollection(firebaseInstance, userId).doc(positionId);
