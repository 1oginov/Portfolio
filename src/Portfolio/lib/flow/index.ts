/* @flow */

export type Position = {
  id: string,
  symbol: string,
  price: number,
  amount: number,
  date: string,
};
