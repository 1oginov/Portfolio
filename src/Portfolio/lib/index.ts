// TODO: Tests.

export interface Position {
  id: string;
  symbol: string;
  amount: number;

  openPrice: number;
  openCommission: number;
  openDate: string;

  closePrice: number | null;
  closeCommission: number | null;
  closeDate: string | null;
}

export const annualizePercent = (percent: number, fromDate: string, toDate?: string): number => {
  if (percent === 0) {
    return 0;
  }

  const from = new Date(fromDate).getTime();
  const to = toDate ? new Date(toDate).getTime() : Date.now();

  if (to === from) {
    return 0;
  }

  const oneYear = 24 * 60 * 60 * 1000 * 365.25;

  const years = (to - from) / oneYear;

  return percent / years;
};

export const formatDate = (date: Date): string => date.toISOString().slice(0, 10);

export const getUniqueSymbolsFromPositions = (positions: Position[]): string[] => Array.from(
  new Set(positions.map(position => position.symbol)),
);
