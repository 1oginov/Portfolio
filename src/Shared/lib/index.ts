export const areArraysEqual = <T>(a: T[], b: T[]): boolean => {
  if (a.length !== b.length) {
    return false;
  }

  const sortedA = a.sort();
  const sortedB = b.sort();

  return sortedA.every((value, index) => value === sortedB[index]);
};

export const firebaseCollectionToArray = <T>(object: { [key: string]: T }): (T & { id: string })[] => (
  Object.keys(object).map(id => ({
    ...object[id],
    id,
  }))
);
