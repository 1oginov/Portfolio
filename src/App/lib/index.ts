import { annualizePercent, Position } from 'Portfolio/lib';
import { getQuotePrice } from 'Stocks/lib';
import Quote from 'Stocks/lib/IEX/Quote';

// TODO: Tests.

export interface StockPosition extends Position {
  openSum: number;

  closeSum: number | null;
  closePL: number | null;
  closePLPercent: number | null;
  closePLAnnualPercent: number | null;

  quote: Quote | null;
  quoteProgress: boolean;

  companyName: string | null;
  dailyPL: number | null;
  dailyPLPercent: number | null;

  marketPrice: number | null;
  marketSum: number | null;
  marketPL: number | null;
  marketPLPercent: number | null;
  marketPLAnnualPercent: number | null;
}

export interface OpenPositionsSummary {
  symbol: string;
  amount: number;

  openPriceAverage: number;
  openSum: number;

  companyName: string | null;
  dailyPL: number | null;
  dailyPLPercent: number | null;

  marketPL: number | null;
  marketPLPercent: number | null;
}

export interface Totals {
  totalOpenSum: number;

  totalCloseSum: number;
  totalClosePL: number;
  totalClosePLPercent: number;

  totalDailyPL: number;
  totalDailyPLPercent: number;

  totalMarketPL: number;
  totalMarketPLPercent: number;
  totalMarketSum: number;
}

export const createStockPosition = (position: Position, quote: Quote | null, quoteProgress: boolean): StockPosition => {
  const openSum = position.amount * position.openPrice + position.openCommission;

  let closeSum = null;
  let closePL = null;
  let closePLPercent = null;
  let closePLAnnualPercent = null;

  if (position.closeCommission !== null && position.closeDate !== null && position.closePrice !== null) {
    closeSum = position.amount * position.closePrice - position.closeCommission;
    closePL = closeSum - openSum;
    closePLPercent = closePL / openSum;
    closePLAnnualPercent = annualizePercent(closePLPercent, position.openDate, position.closeDate);
  }

  let companyName = null;
  let dailyPL = null;
  let dailyPLPercent = null;

  if (quote !== null) {
    ({ companyName } = quote);
    dailyPL = position.amount * quote.change;
    dailyPLPercent = quote.changePercent;
  }

  const marketPrice = getQuotePrice(quote);
  let marketSum = null;
  let marketPL = null;
  let marketPLPercent = null;
  let marketPLAnnualPercent = null;

  if (marketPrice !== null) {
    marketSum = position.amount * marketPrice;
    marketPL = marketSum - openSum;
    marketPLPercent = marketPL / openSum;
    marketPLAnnualPercent = annualizePercent(marketPLPercent, position.openDate);
  }

  return {
    ...position,
    closePL,
    closePLAnnualPercent,
    closePLPercent,
    closeSum,
    companyName,
    dailyPL,
    dailyPLPercent,
    marketPL,
    marketPLAnnualPercent,
    marketPLPercent,
    marketPrice,
    marketSum,
    openSum,
    quote,
    quoteProgress,
  };
};

export const createOpenPositionsSummaries = (stockPositions: StockPosition[]): OpenPositionsSummary[] => {
  const bySymbols: { [key: string]: StockPosition[] } = {};

  stockPositions.forEach(position => {
    if (!bySymbols[position.symbol]) {
      bySymbols[position.symbol] = [];
    }

    bySymbols[position.symbol].push(position);
  });

  return Object.keys(bySymbols).map(symbol => {
    let amount = 0;
    let openPrice = 0;
    let openSum = 0;
    let dailyPL: number | null = null;
    let dailyPLPercent: number | null = null;
    let marketPL: number | null = null;
    let marketPLPercent: number | null = null;

    bySymbols[symbol].forEach(stockPosition => {
      amount += stockPosition.amount;
      openPrice += stockPosition.amount * stockPosition.openPrice;
      openSum += stockPosition.openSum;

      if (stockPosition.dailyPL !== null) {
        if (dailyPL === null) {
          dailyPL = 0;
        }

        dailyPL += stockPosition.dailyPL;
      }

      if (stockPosition.marketPL !== null) {
        if (marketPL === null) {
          marketPL = 0;
        }

        marketPL += stockPosition.marketPL;
      }
    });

    const referencePosition = bySymbols[symbol][0];

    if (dailyPL !== null && referencePosition.quote !== null) {
      dailyPLPercent = dailyPL / amount / referencePosition.quote.close;
    }

    if (marketPL !== null) {
      marketPLPercent = marketPL / openSum;
    }

    const openPriceAverage = openPrice / amount;
    const { companyName } = referencePosition;

    return {
      amount,
      companyName,
      dailyPL,
      dailyPLPercent,
      marketPL,
      marketPLPercent,
      openPriceAverage,
      openSum,
      symbol,
    };
  });
};

export const calculateTotals = (stockPositions: StockPosition[]): Totals => {
  let totalOpenSum = 0;
  let totalCloseSum = 0;
  let totalDailyPL = 0;
  let totalPreviousCloseSum = 0;
  let totalMarketSum = 0;

  stockPositions.forEach(position => {
    totalOpenSum += position.openSum;

    if (position.closeSum !== null) {
      totalCloseSum += position.closeSum;
    }

    if (position.quote !== null) {
      totalDailyPL += position.amount * position.quote.change;
      totalPreviousCloseSum += position.amount * position.quote.close;
    }

    if (position.marketSum !== null) {
      totalMarketSum += position.marketSum;
    }
  });

  const totalClosePL = totalCloseSum - totalOpenSum;
  const totalClosePLPercent = totalClosePL / totalOpenSum;

  const totalDailyPLPercent = totalDailyPL / totalPreviousCloseSum;

  const totalMarketPL = totalMarketSum - totalOpenSum;
  const totalMarketPLPercent = totalMarketPL / totalOpenSum;

  return {
    totalClosePL,
    totalClosePLPercent,
    totalCloseSum,
    totalDailyPL,
    totalDailyPLPercent,
    totalMarketPL,
    totalMarketPLPercent,
    totalMarketSum,
    totalOpenSum,
  };
};
