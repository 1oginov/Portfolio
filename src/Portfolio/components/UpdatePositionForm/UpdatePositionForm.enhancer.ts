import * as React from 'react';
import { connect } from 'react-redux';
import {
  compose, StateHandler, withHandlers, withStateHandlers,
} from 'recompose';

import { formatDate } from 'Shared/lib';

import { updatePosition as updatePositionAction, UpdatePositionAction } from '../../actions';
import { Position } from '../../lib';
import { Props } from './UpdatePositionForm';

const mapDispatchToProps = { updatePosition: updatePositionAction };

interface DispatchProps {
  updatePosition: UpdatePositionAction;
}

interface WithStateHandlersState {
  commission: number | '';
  date: string;
  price: number | '';
}

interface WithStateHandlersUpdaters {
  [key: string]: StateHandler<WithStateHandlersState>;
}

interface EnhancedProps {
  backButton?: React.ReactNode;
  id: string;
  onClose?: () => void;
  position: Position;
}

export default compose<Props & DispatchProps & WithStateHandlersState, EnhancedProps>(
  connect(null, mapDispatchToProps),
  withStateHandlers<WithStateHandlersState, WithStateHandlersUpdaters>(
    {
      commission: '',
      date: formatDate(new Date()),
      price: '',
    },
    {
      handleCommissionChange: () => event => {
        const commission = parseFloat(event.target.value);

        return {
          commission: commission >= 0 ? commission : '',
        };
      },
      handleDateChange: () => event => {
        let date;

        try {
          date = formatDate(new Date(event.target.value));
        } catch (error) {
          //
        }

        if (!date) {
          date = formatDate(new Date());
        }

        return { date };
      },
      handlePriceChange: () => event => {
        const price = parseFloat(event.target.value);

        return {
          price: price >= 0 ? price : '',
        };
      },
    },
  ),
  withHandlers<EnhancedProps & DispatchProps & WithStateHandlersState & WithStateHandlersUpdaters, {}>({

    handleSubmit: ({
      commission, date, onClose, position, price, updatePosition,
    }) => (event: React.SyntheticEvent) => {
      event.preventDefault();

      if (commission === '' || price === '') {
        // TODO: Display the following error in UI.
        throw new Error('Invalid values');
      }

      updatePosition({
        ...position, closeCommission: commission, closeDate: date, closePrice: price,
      })
        .then(() => {
          if (onClose) {
            onClose();
          }
        });
    },

  }),
);
