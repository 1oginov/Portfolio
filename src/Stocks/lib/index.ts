import { Repository } from 'redux-repository/lib/interfaces';
import { getResourceById } from 'redux-repository/lib/repository';
import { extractData } from 'redux-repository/lib/resource';

import Quote from './IEX/Quote';

export const findQuoteBySymbol = (quotesRepository: Repository<Quote, string>, symbol: string): Quote | null => (
  extractData(getResourceById(quotesRepository, symbol))
);

export const getQuotePrice = (quote: Quote | null): number | null => {
  if (quote) {
    if (quote.iexRealtimePrice) {
      return quote.iexRealtimePrice;
    }

    return quote.latestPrice;
  }

  return null;
};
