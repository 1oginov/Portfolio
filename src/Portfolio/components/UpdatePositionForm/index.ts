import UpdatePositionForm from './UpdatePositionForm';
import enhance from './UpdatePositionForm.enhancer';
import style from './UpdatePositionForm.styles';

export default style(enhance(UpdatePositionForm));
