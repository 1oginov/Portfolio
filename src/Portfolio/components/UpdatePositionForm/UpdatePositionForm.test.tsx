import { shallow } from 'enzyme';
import classes from 'identity-obj-proxy';
import * as React from 'react';

import { formatDate } from 'Shared/lib';

import UpdatePositionForm from './UpdatePositionForm';

it('matches snapshot', () => {
  const wrapper = shallow(
    <UpdatePositionForm
      classes={classes}
      commission={1}
      date={formatDate(new Date('2019-07-02'))}
      handleCommissionChange={() => undefined}
      handleDateChange={() => undefined}
      handlePriceChange={() => undefined}
      handleSubmit={() => undefined}
      price={100}
    />,
  );

  expect(wrapper).toMatchSnapshot();
});
