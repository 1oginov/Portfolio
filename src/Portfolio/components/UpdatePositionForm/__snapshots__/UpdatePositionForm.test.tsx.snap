// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<form
  className="root"
  onSubmit={[Function]}
>
  <div
    className="row"
  >
    <WithStyles(ForwardRef(TextField))
      fullWidth={true}
      inputProps={
        Object {
          "min": 0,
          "step": 0.01,
        }
      }
      label="Price"
      onChange={[Function]}
      required={true}
      type="number"
      value={100}
      variant="outlined"
    />
  </div>
  <div
    className="row"
  >
    <WithStyles(ForwardRef(TextField))
      fullWidth={true}
      inputProps={
        Object {
          "min": 0,
          "step": 0.01,
        }
      }
      label="Commission"
      onChange={[Function]}
      required={true}
      type="number"
      value={1}
      variant="outlined"
    />
  </div>
  <div
    className="row"
  >
    <WithStyles(ForwardRef(TextField))
      fullWidth={true}
      label="Date"
      onChange={[Function]}
      required={true}
      type="date"
      value="2019-07-02"
      variant="outlined"
    />
  </div>
  <div
    className="row"
  >
    <WithStyles(ForwardRef(Button))
      color="primary"
      type="submit"
      variant="contained"
    >
      Update
    </WithStyles(ForwardRef(Button))>
  </div>
</form>
`;
